internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,84 @@
 package asset
 
 import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
 	"fmt"
 	"io"
 	"io/fs"
 	"path/filepath"
 
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
 	"github.com/spf13/afero"
+	"golang.org/x/sync/errgroup"
 )
 
-// UploadFunc is the function signature to upload contents to a destination.
-type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+// CacheUploader uploads local asset files.
+type CacheUploader struct {
```

Comment:
What do you think of renaming to `ArtifactBucketUploader` to highlight the fact that this uploader is strictly to put assets to our artifact bucket (vs. maybe a customer's existing bucket)